### ISSUE TITLE
Timeout while waiting for handshake or handshake errors

### ISSUE THREAD

**Issue by munchj** [theophilusx/ssh2-sftp-client#384]
Hello,

I was working on a script to fetch some files from an sftp server through a SOCKS5 proxy and I noticed some strange behaviour.
I managed to use a workaround to make it work, but I saw that people are experiencing this issue so it might be worth investigating more.

I was getting timeouts (Only when using the proxy though) when calling the sftp.connect(ssh_config) function.
So I tried using the ssh2 library directly, and I was able to connect without any issue (through the proxy)
My workaround was to connect the ssh2 client from outside the ssh2-sftp-client library and do the same calls as the ssh2-sftp-client library afterwards.

```js
const SSHClient = require('ssh2').Client;
const sftpClient = require("ssh2-sftp-client");

const sftp = new sftpClient();

var ssh_config = {
    host: process.env.ssh_hostname,
    port: process.env.ssh_port,
    username: process.env.ssh_username,
    privateKey: process.env.ssh_key_secret,
    passphrase: process.env.ssh_key_passphrase,
    debug: (txt) => { console.log('[ssh]', txt) },
    sock: socket
};

function getSftpConnection() {
    return new Promise((resolve) => {
        var conn = new SSHClient();
        sftp.client = conn;
        conn.connect(ssh_config);
        conn.on('ready', function () {
            console.log('sftp connection ready !');
            resolve(conn);
        });
    });
}

await getSftpConnection();
await sftp.getSftpChannel();
await sftp.get(
    files_config.remote.inputFilePath,
    files_config.local.inputFilePath
); //Works !
```

The original solution that is failing is :
```js
const sftpClient = require("ssh2-sftp-client");
const sftp = new sftpClient();

var ssh_config = {
    host: process.env.ssh_hostname,
    port: process.env.ssh_port,
    username: process.env.ssh_username,
    privateKey: process.env.ssh_key_secret,
    passphrase: process.env.ssh_key_passphrase,
    debug: (txt) => { console.log('[ssh]', txt) },
    sock: socket
};

await sftp.connect(ssh_config); //Fails !
await sftp.get(
    files_config.remote.inputFilePath,
    files_config.local.inputFilePath
);
```

I tried to find the cause but didn't see anything obvious. Maybe it has something to do with one of the event listeners initialized by the library.

**Comment by theophilusx** [comment]
I doubt listeners are the issue. The only difference between ssh2 and
ssh2-sftp-client connect is that ssh2-sftp-client's connect method is
wrapped inside a retry library which will cause the connect to retry
multiple times when a connection fails.

Can you do the following?

Tell me what version of ssh2-sftp-client and ssh2 you are using.

Tell me what platform you are running on.

Provide the debug output for both your working version and the
non-working version.

Try using the ssh2-sftp-client HEAD version from the repository. There
are some fixes I've been waiting to push out. While not directly
related, it would be good to verify the issue still exists with the most
recent code. (I've been waiting for a new ssh2 release as there are some
known issues in ssh2 which have been fixed, but not yet released).

**Comment by munchj** [comment]
Thank you for the quick reply !

ssh2-sftp-client : 7.2.1
ssh2 : 1.5.0

The script is running on heroku (linux containers)

working debug logs : 
```
[ssh] Custom crypto binding available   
[ssh] Local ident: 'SSH-2.0-ssh2js1.5.0'
[ssh] Socket connected
[ssh] Remote ident: 'SSH-2.0-OpenSSH_7.4p1 Debian-10+deb9u7'
[ssh] Outbound: Sending KEXINIT
[ssh] Inbound: Handshake in progress
[ssh] Handshake: (local) KEX method: curve25519-sha256@libssh.org,curve25519-sha256,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512
[ssh] Handshake: (remote) KEX method: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1
[ssh] Handshake: KEX algorithm: curve25519-sha256@libssh.org
[ssh] Handshake: (local) Host key format: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256,ssh-rsa
[ssh] Handshake: (remote) Host key format: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ecdsa-sha2-nistp256,ssh-ed25519
[ssh] Handshake: Host key format: ssh-ed25519
[ssh] Handshake: (local) C->S cipher: aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-ctr,aes192-ctr,chacha20-poly1305@openssh.com,aes256-ctr
[ssh] Handshake: (remote) C->S cipher: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
[ssh] Handshake: C->S Cipher: aes128-gcm@openssh.com
[ssh] Handshake: (local) S->C cipher: aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-ctr,aes192-ctr,chacha20-poly1305@openssh.com,aes256-ctr
[ssh] Handshake: (remote) S->C cipher: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
[ssh] Handshake: S->C cipher: aes128-gcm@openssh.com
[ssh] Handshake: (local) C->S MAC: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
[ssh] Handshake: (remote) C->S MAC: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
[ssh] Handshake: C->S MAC: <implicit>
[ssh] Handshake: (local) S->C MAC: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
[ssh] Handshake: (remote) S->C MAC: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
[ssh] Handshake: S->C MAC: <implicit>
[ssh] Handshake: (local) C->S compression: none,zlib@openssh.com,zlib
[ssh] Handshake: (remote) C->S compression: none,zlib@openssh.com
[ssh] Handshake: C->S compression: none
[ssh] Handshake: (local) S->C compression: none,zlib@openssh.com,zlib
[ssh] Handshake: (remote) S->C compression: none,zlib@openssh.com
[ssh] Handshake: S->C compression: none
[ssh] Outbound: Sending KEXECDH_INIT
[ssh] Received DH Reply
[ssh] Host accepted by default (no verification)
[ssh] Host accepted (verified)
[ssh] Outbound: Sending NEWKEYS
[ssh] Inbound: NEWKEYS
[ssh] Verifying signature ...
[ssh] Verified signature
[ssh] Handshake completed
[ssh] Outbound: Sending SERVICE_REQUEST (ssh-userauth)
[ssh] Inbound: Received SERVICE_ACCEPT (ssh-userauth)
[ssh] Outbound: Sending USERAUTH_REQUEST (none)
[ssh] Inbound: Received USERAUTH_FAILURE (publickey,password)
[ssh] Client: none auth failed
[ssh] Outbound: Sending USERAUTH_REQUEST (publickey -- check)
[ssh] Inbound: Received USERAUTH_PK_OK
[ssh] Outbound: Sending USERAUTH_REQUEST (publickey)
[ssh] Inbound: Received USERAUTH_SUCCESS
[ssh] Outbound: Sending CHANNEL_OPEN (r:0, session)
[ssh] Inbound: GLOBAL_REQUEST (hostkeys-00@openssh.com)
[ssh] Outbound: Sending GLOBAL_REQUEST (hostkeys-prove-00@openssh.com)
[ssh] Inbound: CHANNEL_OPEN_CONFIRMATION (r:0, s:0)
[ssh] Outbound: Sending CHANNEL_REQUEST (r:0, subsystem: sftp)
[ssh] Inbound: REQUEST_SUCCESS
[ssh] Inbound: CHANNEL_WINDOW_ADJUST (r:0, 2097152)
[ssh] Inbound: CHANNEL_SUCCESS (r:0)
[ssh] Outbound: Sending CHANNEL_DATA (r:0, 9)
[ssh] Inbound: CHANNEL_DATA (r:0, 154)
[ssh] SFTP: Inbound: Received VERSION (v3, exts:posix-rename@openssh.com,statvfs@openssh.com,fstatvfs@openssh.com,hardlink@openssh.com,fsync@openssh.com)
```

non working solution : 
```
[ssh] CLIENT[sftp]: connect: Debugging turned on
[ssh] CLIENT[sftp]: connect: Connect attempt 1
[ssh] CLIENT[sftp]: getConnection: Adding temp event listeners
[ssh] CLIENT[sftp]: getConnection: created promise
[ssh] CLIENT[sftp]: Adding listener to ready event
[ssh] Custom crypto binding available
[ssh] Local ident: 'SSH-2.0-ssh2js1.5.0'
[ssh] Socket connected
[ssh] CLIENT[sftp]: getConnection: Handling error: Timed out while waiting for handshake
[ssh] CLIENT[sftp]: getConnection: handled error with reject
[ssh] CLIENT[sftp]: Global: Ignoring handled error
[ssh] CLIENT[sftp]: getConnection: finally clause fired
[ssh] Socket closed
[ssh] CLIENT[sftp]: getConnection: Ignoring handled error: Connection lost before handshake
[ssh] CLIENT[sftp]: Global: Ignoring handled error
[ssh] CLIENT[sftp]: getConnection: handling unexpected close event
[ssh] CLIENT[sftp]: getConnection: handling close event with reject
[ssh] CLIENT[sftp]: Global: Ignoring handled close event
[ssh] CLIENT[sftp]: getConnection: Removing temp event listeners
[ssh] CLIENT[sftp]: getConnection retry catch
[ssh] CLIENT[sftp]: getConnection: Adding temp event listeners
[ssh] CLIENT[sftp]: getConnection: created promise
[ssh] CLIENT[sftp]: Adding listener to ready event
[ssh] Custom crypto binding available
[ssh] Local ident: 'SSH-2.0-ssh2js1.5.0'
[ssh] Socket connected
[ssh] CLIENT[sftp]: getConnection: Handling error: Timed out while waiting for handshake
[ssh] CLIENT[sftp]: getConnection: handled error with reject
[ssh] CLIENT[sftp]: Global: Ignoring handled error
[ssh] CLIENT[sftp]: getConnection: finally clause fired
[ssh] CLIENT[sftp]: Removing listener from ready event
[ssh] CLIENT[sftp]: getConnection: Removing temp event listeners
[ssh] CLIENT[sftp]: getConnection retry catch
[ssh] CLIENT[sftp]: connect: Error getConnection: Timed out while waiting for handshake
script execution failed Error: connect->getConnection: Timed out while waiting for handshake
```

I guess that I'm dependent on npm on heroku but when I have some time I will test locally

**Comment by theophilusx** [comment]
Thanks for the logs. I'll go through them and see if I can notice
something. I'd like to avoid having to setup a socks proxy if possible,
but may have to in order to reproduce and diagnose. This could take some
time.

Note that you should be able to replace the npm version with the repo
version by replacing the version number with
theophilusx/ssh2-sftp-client

i.e.

dependencies {
  ...
  "ssh2-sftp-client": "theophilusx/ssh2-sftp-client",
  ...
}

You will need to remove node_modules and package-lock.json and then do
an npm i to re-install.

From a quick scan of the logs, it does look like the initial connection
is made, but then a timeout exception is being thrown for some reason.
It is possible that what is happening is that the retry module is timing
out before the slower connection via a proxy can complete. As the ssh2
library does not have any retry facility, it is not affected by this.
However, that is just a guess at this stage. You could try adding a
minTimeout property to the config object. This will set how long the
retry module waits before attempting another connection. Default is
1000ms. Perhaps try setting this to something like 4000 and then if that
works, slowly reduce it until you find a timeout which works.

**Comment by munchj** [comment]
I just tried with 
```
    "ssh2-sftp-client": {
      "version": "github:theophilusx/ssh2-sftp-client#8cefc463047cb41b168a71bdf67c72d7d915a4e9",
      "from": "github:theophilusx/ssh2-sftp-client",
      "requires": {
        "concat-stream": "^2.0.0",
        "promise-retry": "^2.0.1",
        "ssh2": "^1.5.0"
      }
    },
```

and good news, it seems to work !

**Comment by theophilusx** [comment]
OK. I have been waiting for a new ssh2 version before pushing out the
new ssh2-sftp-client version, but I think I will just have to go ahead
and push out a new one now and do another one later if necessary once a
new ssh2 is released. Bit of a pain as I know there are a couple of
issues people are encountering which are fixed in ssh2 master.

I'll push a new version out on Monday.

**Comment by theophilusx** [comment]
New version of ssh2 just released, so pushing out 7.2.2 now. Will take a little time to get to all the npmjs.com servers.
